subprojects/files/src/main/java/org/gradle/internal/file/impl/DefaultDeleter.java

```diff
@@ -152,29 +158,40 @@ private boolean shouldRemoveContentsOf(File file, Handling handling) {
         return file.isDirectory() && (handling.shouldFollowLinkedDirectory() || !isSymlink.test(file));
     }
 
-    protected boolean deleteFile(File file) {
-        return file.delete() && !file.exists();
+    protected boolean deleteFile(File file) throws IOException {
+        Path path = file.toPath();
+        return Files.deleteIfExists(path) && !Files.exists(path);
     }
 
     private boolean tryHardToDelete(File file) {
-        if (deleteFile(file)) {
-            return true;
-        }
-
-        // This is copied from Ant (see org.apache.tools.ant.util.FileUtils.tryHardToDelete).
-        // It mentions that there is a bug in the Windows JDK implementations that this is a valid
-        // workaround for. I've been unable to find a definitive reference to this bug.
-        // The thinking is that if this is good enough for Ant, it's good enough for us.
-        if (runGcOnFailedDelete) {
-            System.gc();
-        }
+        boolean result = false;
         try {
-            Thread.sleep(DELETE_RETRY_SLEEP_MILLIS);
-        } catch (InterruptedException ex) {
-            Thread.currentThread().interrupt();
+            result = deleteFile(file);
+        } catch (DirectoryNotEmptyException exception) {
+            int failedAttempts = 1;
+            while (runRetryOnFailedDelete && failedAttempts < EMPTY_DIRECTORY_DELETION_ATTEMPTS) {
+                try {
+                    Thread.sleep(DELETE_RETRY_SLEEP_MILLIS);
+                } catch (InterruptedException ex) {
+                    Thread.currentThread().interrupt();
+                }
+                try {
+                    result = deleteFile(file);
+                } catch (DirectoryNotEmptyException ex) {
+                    failedAttempts++;
+                    continue;
+                } catch (Exception ex) {
```

Comment:
❓ I would only catch `IOException` here to let unexpected errors pop up

Reply:
Done